keps/sig-network/2610-allports-services/README.md

```diff
@@ -0,0 +1,553 @@
+This KEP proposes to add a new field to the ServicePort spec to allow exposing the entire port range(1 to 65535).
+
+## Motivation
+
+There are several applications like SIP apps or RTP which needs a lot of ports to run multiple calls or media streams.
+Currently the only option is to specify every port in the Service Spec. A request for port ranges in Services has been open in https://github.com/kubernetes/kubernetes/issues/23864. Implementing port ranges are challenging since iptables/ipvs do not support remapping port ranges. Hence, the proposal to set a single field in order to expose the entire port range and implement the service clients and endpoints accordingly.
+
+### Goals
+
+* Allow users to optionally expose the entire port range via a Service (of Type LoadBalancer or ClusterIP).
+
+### Non-Goals
+
+* Supporting Port Ranges in a Service.
+* Changing the default behavior of Service ports.
+
+## Proposal
+
+The proposal here is to introduce an “allPorts” boolean field to the [servicePort API.](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#serviceport-v1-core)
+This field is introduced within ServicePort rather than ServiceSpec, because a ServicePort entry is needed to infer the protocol.
+
+This field will be supported for Service Type=ClusterIP and ServiceType=LoadBalancer. This is not applicable to ExternalName services. NodePort services are not supported.
+A NodePort service accepts traffic on a given port on the Node, redirecting it to the specified targetPort of the service endpoints. Support AllPorts for a NodePort service means -
+traffic to any port on the node, will be forwarded to its endpoints on the same port. This could potentially break networking on the node, if traffic for, say, port 22 got forwarded to \<endpoint IP\>:22.
+
+In order to expose the entire port range on a supported service, a user needs to:
+1) Create a single ServicePort entry in the ServiceSpec and set "allPorts" to True.
+2) Populate the protocol field.
+3) Populate a valid port in the port field.
+   This is required to make sure we do not break other Service API clients (like DNS) that might always expect a non-zero port value. The [APIServer validation](https://github.com/kubernetes/kubernetes/blob/b9ce4ac212d150212485fa29d62a2fbd783a57b0/pkg/apis/core/validation/validation.go#L4373) could be relaxed, but we might break other clients.
+
+It is possible to specify more than 1 port in the serviceSpec when using the [Mixed Protocol Feature](https://kubernetes.io/docs/concepts/services-networking/service/#load-balancers-with-mixed-protocol-types). There will one ServicePort object per protocol.
+By default, a NodePort will assigned for each valid port(one per protocol) in the service, if it is of Service Type=LoadBalancer.
+Traffic Forwarding will work as usual for those NodePorts.
+NodePort allocation can be disabled by setting "allocateLoadBalancerNodePorts: false" on the service spec.
```

Comment:
I think `allocateLoadBalancerNodePorts: false` will only prevent _automatic_ allocation of nodePorts. If you define an explicit nodePort it will be allocated.

Reply:
I agree.. the multiple restrictions are confusing. I think the next step here is to 1) Edit this KEP with the LoadBalancer+ Headless service proposal. 2) Any bigger changes should probably be part of Gateway L4.
I will refactor this KEP.

Reply:
I have refactored the KEP and marking this as resolved.